src/Migrations/HSLMigration.php

```diff
@@ -0,0 +1,684 @@
+          $new_length = new ListItem(
+            new LiteralExpression(new DecimalLiteralToken(
+              HHAST\Missing(),
+              HHAST\Missing(),
+              (string)$rewrite_length_value,
+            )),
+            HHAST\Missing(),
+          );
+        } else {
+          // with a positive offset this is harder
+          // we need to replace this arg with a more complex expression
+          // based on the length of the string
+          $haystack = $items[0]->getCode();
+          $new_length = $this->nodeFromCode(
+            'Str\\length('.$haystack.') - '.($offset + Math\abs($length)),
+            ExpressionStatement::class,
+          );
+        }
+
+        // rewrite args list
+        $new_argument_list = $argument_list->replace($items[2], $new_length);
+      }
+    } elseif ($fn_name === 'Str\\splice' && C\count($items) === 4) {
+      // check for negative length arguments to Str\splice, which will throw a runtime exception
+      // this is currently unhandled, so we just bail by returning null if we find it
+      $length = $this->resolveIntegerArgument($items[3]);
+      if ($length !== null && $length < 0) {
+        return null;
+      }
+    } elseif (
+      ($fn_name === 'Math\\max' || $fn_name === 'Math\\min') &&
+      C\count($items) !== 1
+    ) {
+      // PHP max() and min() either take a list of variadic args, or an array of args
+      // in HSL, max and min want a single Traversable arg, while maxva and minva are variadic
+      return $this->replaceFunctionName($node, $fn_name.'va');
+    } elseif ($fn_name === 'Math\\round' && C\count($items) > 2) {
+      // can't handle the optional third argument of round()
+      return null;
+    }
+
+    if ($argument_order !== null) {
+      $new_items = vec[];
+      foreach ($argument_order as $index) {
+        $new_items[] = $items[$index];
+      }
+
+      $new_argument_list = vec[];
+
+      foreach ($arguments as $i => $argument) {
+        invariant($argument instanceof ListItem, 'expected ListItem');
+        $new_argument_list[] =
+          $argument->replace($argument->getItem(), $new_items[(int)$i]);
+      }
+
+      $new_argument_list = EditableList::fromItems($new_argument_list);
+    }
+
+    return $node->replace($argument_list, $new_argument_list);
+  }
+
+  // many PHP functions can return false, and their HSL counterparts return null instead
+  // this will replace false with null in binary expressions like === false and !=== false
+  protected function maybeChangeFalseToNull(
+    EditableNode $root,
+    FunctionCallExpression $node,
+  ): EditableNode {
+    $parents = null;
+    $found = false;
+    $stack = $root->findWithParents($it ==> $it === $node);
+    invariant(!C\is_empty($stack), 'did not find node in root');
+    invariant(C\lastx($stack) === $node, 'expected node at top of stack');
+    $stack_count = C\count($stack);
+    $parent = $stack[$stack_count - 2];
+
+    if ($parent instanceof BinaryExpression) {
```

Comment:
this also looks like a good candidate for negate the check, return early, unindent the rest